src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -141,45 +141,43 @@ private static void IntroSort(Span<T> keys, int depthLimit, Comparison<T> compar
             Debug.Assert(depthLimit >= 0);
             Debug.Assert(comparer != null);
 
-            int hi = keys.Length - 1;
-            while (hi > 0)
+            int partitionSize = keys.Length;
+            while (partitionSize > 1)
             {
-                int partitionSize = hi + 1;
-
                 if (partitionSize <= Array.IntrosortSizeThreshold)
                 {
                     Debug.Assert(partitionSize >= 2);
```

Comment:
@stephentoub probably could remove the debug.assert's after this change, since it is a no-brainer now :)
```suggestion
```